File: blue-core/src/main/java/blue/time/TimeUnitMath.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2026 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */
package blue.time;

/**
 * Type-safe arithmetic operations between {@link TimePosition} (position) and
 * {@link TimeDuration} (duration).
 * 
 * Method signatures enforce valid operations at compile time:
 * <ul>
 *   <li>Position + Duration → Position</li>
 *   <li>Position − Position → Duration (via {@link #distance})</li>
 *   <li>Duration + Duration → Duration</li>
 *   <li>Duration − Duration → Duration</li>
 * </ul>
 * 
 * Inspired by Ardour's deliberate disabling of {@code operator-} on
 * {@code timepos_t} to prevent semantically ambiguous operations.
 * 
 * All operations route through Csound beats as the canonical intermediate.
 * The result preserves the TimeBase of the first operand where applicable.
 *
 * @author Steven Yi
 */
public final class TimeUnitMath {
    
    private TimeUnitMath() {} // utility class — not instantiable
    
    // ========== Position + Duration → Position ==========
    
    /**
     * Adds a duration to a position, returning a new position.
     * The result preserves the TimeBase of the position operand.
     * 
     * @param context the TimeContext for conversion
     * @param position the starting position
     * @param duration the duration to add
     * @return a new TimePosition representing position + duration
     */
    public static TimePosition add(TimeContext context, TimePosition position, TimeDuration duration) {
        double posBeats = position.toBeats(context);
        double durBeats = duration.toBeats(context);
        double resultBeats = posBeats + durBeats;
        return TimeUtilities.beatsToTimePosition(resultBeats, position.getTimeBase(), context);
    }
    
    // ========== Position − Position → Duration ==========
    
    /**
     * Computes the distance between two positions as a duration.
     * Returns the absolute distance (always non-negative).
     * The result is in DurationBeats.
     * 
     * @param context the TimeContext for conversion
     * @param from the starting position
     * @param to the ending position
     * @return a TimeDuration representing the distance between the positions
     */
    public static TimeDuration distance(TimeContext context, TimePosition from, TimePosition to) {
        double fromBeats = from.toBeats(context);
        double toBeats = to.toBeats(context);
        return TimeDuration.beats(Math.abs(toBeats - fromBeats));
    }
    
    /**
     * Computes the signed distance from one position to another.
     * Positive if {@code to} is after {@code from}, negative is not possible
     * since TimeDuration cannot be negative — use {@link #distance} instead.
     * Returns zero if {@code to} is before {@code from}.
     * 
     * @param context the TimeContext for conversion
     * @param from the starting position
     * @param to the ending position
     * @return a TimeDuration representing max(0, to - from)
     */
    public static TimeDuration forwardDistance(TimeContext context, TimePosition from, TimePosition to) {
        double fromBeats = from.toBeats(context);
        double toBeats = to.toBeats(context);
        return TimeDuration.beats(Math.max(0, toBeats - fromBeats));
    }
    
    // ========== Duration + Duration → Duration ==========
    
    /**
     * Adds two durations together.
     * The result is in DurationBeats.
     * 
     * @param context the TimeContext for conversion
     * @param a the first duration
     * @param b the second duration
     * @return a TimeDuration representing a + b
     */
    public static TimeDuration add(TimeContext context, TimeDuration a, TimeDuration b) {
        double aBeats = a.toBeats(context);
        double bBeats = b.toBeats(context);
        return TimeDuration.beats(aBeats + bBeats);
    }
    
    // ========== Duration − Duration → Duration ==========
    
    /**
     * Subtracts one duration from another.
     * The result is clamped to zero (never negative).
     * The result is in DurationBeats.
     * 
     * @param context the TimeContext for conversion
     * @param a the duration to subtract from
     * @param b the duration to subtract
     * @return a TimeDuration representing max(0, a - b)
     */
    public static TimeDuration subtract(TimeContext context, TimeDuration a, TimeDuration b) {
        double aBeats = a.toBeats(context);
        double bBeats = b.toBeats(context);
        return TimeDuration.beats(Math.max(0, aBeats - bBeats));
    }
    
    // ========== Position − Duration → Position ==========
    
    /**
     * Subtracts a duration from a position, returning a new position.
     * The result is clamped so the position never goes below zero beats.
     * The result preserves the TimeBase of the position operand.
     * 
     * @param context the TimeContext for conversion
     * @param position the starting position
     * @param duration the duration to subtract
     * @return a new TimePosition representing max(0, position - duration)
     */
    public static TimePosition subtract(TimeContext context, TimePosition position, TimeDuration duration) {
        double posBeats = position.toBeats(context);
        double durBeats = duration.toBeats(context);
        double resultBeats = Math.max(0, posBeats - durBeats);
        return TimeUtilities.beatsToTimePosition(resultBeats, position.getTimeBase(), context);
    }
    
    // ========== Conversion Bridges ==========
    
    /**
     * Converts a TimeDuration to a specific TimeBase representation.
     * For measure-based formats (BBT, BBST, BBF), the result uses 0-based
     * duration semantics.
     * 
     * @param duration the duration to convert
     * @param targetTimeBase the desired TimeBase
     * @param context the TimeContext for conversion
     * @return a new TimeDuration in the target TimeBase
     */
    public static TimeDuration convertDuration(TimeDuration duration, TimeBase targetTimeBase, TimeContext context) {
        if (duration.getTimeBase() == targetTimeBase) {
            return duration;
        }
        
        double beats = duration.toBeats(context);
        return beatsToDuration(beats, targetTimeBase, context);
    }
    
    /**
     * Converts Csound beats to a TimeDuration of the specified TimeBase.
     * For measure-based formats, uses 0-based bars/beats (duration semantics).
     * 
     * @param beats the duration in Csound beats
     * @param targetTimeBase the desired TimeBase
     * @param context the TimeContext for conversion
     * @return a new TimeDuration of the specified type
     */
    public static TimeDuration beatsToDuration(double beats, TimeBase targetTimeBase, TimeContext context) {
        if (beats < 0) {
            beats = 0;
        }
        
        return switch (targetTimeBase) {
            case BEATS -> TimeDuration.beats(beats);
            case BBT -> {
                var meter = context.getMeterMap().get(0).getMeter();
                double beatsPerBar = meter.getMeasureBeatDuration();
                double beatScale = 4.0 / meter.beatLength;
                long bars = (long) (beats / beatsPerBar);
                double remainingBeats = beats - (bars * beatsPerBar);
                double scaledBeat = remainingBeats / beatScale;
                int wholeBeat = (int) scaledBeat;
                double fractionalBeat = scaledBeat - wholeBeat;
                int ticks = (int) Math.round(fractionalBeat * TimeContext.DEFAULT_PPQ);
                yield TimeDuration.bbt(bars, wholeBeat, ticks);
            }
            case BBST -> {
                var meter = context.getMeterMap().get(0).getMeter();
                double beatsPerBar = meter.getMeasureBeatDuration();
                double beatScale = 4.0 / meter.beatLength;
                long bars = (long) (beats / beatsPerBar);
                double remainingBeats = beats - (bars * beatsPerBar);
                double scaledBeat = remainingBeats / beatScale;
                int wholeBeat = (int) scaledBeat;
                double fractionalBeat = scaledBeat - wholeBeat;
                int totalTicks = (int) Math.round(fractionalBeat * TimeContext.DEFAULT_PPQ);
                int ticksPerSixteenth = TimeContext.DEFAULT_PPQ / 4;
                int sixteenth = totalTicks / ticksPerSixteenth;
                int subTicks = totalTicks % ticksPerSixteenth;
                yield TimeDuration.bbst(bars, wholeBeat, sixteenth, subTicks);
            }
            case BBF -> {
                var meter = context.getMeterMap().get(0).getMeter();
                double beatsPerBar = meter.getMeasureBeatDuration();
                double beatScale = 4.0 / meter.beatLength;
                long bars = (long) (beats / beatsPerBar);
                double remainingBeats = beats - (bars * beatsPerBar);
                double scaledBeat = remainingBeats / beatScale;
                int wholeBeat = (int) scaledBeat;
                double fractionalBeat = scaledBeat - wholeBeat;
                int fraction = (int) Math.round(fractionalBeat * 100);
                if (fraction >= 100) {
                    fraction = 99;
                }
                yield TimeDuration.bbf(bars, wholeBeat, fraction);
            }
            case TIME -> {
                double seconds = context.getTempoMap().beatsToSeconds(beats);
                long totalMs = Math.round(seconds * 1000);
                long h = totalMs / 3600000;
                long m = (totalMs % 3600000) / 60000;
                long s = (totalMs % 60000) / 1000;
                long ms = totalMs % 1000;
                yield TimeDuration.time(h, m, s, ms);
            }
            case SMPTE -> {
                // SMPTE is display-only — store as DurationTime instead
                double seconds = context.getTempoMap().beatsToSeconds(beats);
                long totalMs = Math.round(seconds * 1000);
                long h = totalMs / 3600000;
                long m = (totalMs % 3600000) / 60000;
                long s = (totalMs % 60000) / 1000;
                long ms = totalMs % 1000;
                yield TimeDuration.time(h, m, s, ms);
            }
            case FRAME -> {
                double seconds = context.getTempoMap().beatsToSeconds(beats);
                long frameCount = Math.round(seconds * context.getSampleRate());
                yield TimeDuration.frames(frameCount);
            }
        };
    }
    
    /**
     * Creates a TimeDuration from a TimePosition value, interpreting the TimePosition's
     * beat value as a duration rather than a position.
     * 
     * This is useful for bridging existing code that stores durations as TimePosition.
     * The resulting TimeDuration will be in DurationBeats format.
     * 
     * @param timePosition the TimePosition whose beat value represents a duration
     * @param context the TimeContext for conversion
     * @return a TimeDuration equivalent
     */
    public static TimeDuration fromTimePosition(TimePosition timePosition, TimeContext context) {
        double beats = timePosition.toBeats(context);
        return TimeDuration.beats(Math.max(0, beats));
    }
    
    /**
     * Creates a TimeDuration from a TimePosition value, converting to the specified
     * TimeBase with proper duration semantics (0-based for measure formats).
     * 
     * @param timePosition the TimePosition whose beat value represents a duration
     * @param targetTimeBase the desired TimeBase for the result
     * @param context the TimeContext for conversion
     * @return a TimeDuration in the target TimeBase
     */
    public static TimeDuration fromTimePosition(TimePosition timePosition, TimeBase targetTimeBase, TimeContext context) {
        double beats = timePosition.toBeats(context);
        return beatsToDuration(Math.max(0, beats), targetTimeBase, context);
    }
}

File: packages/blue-data/src/time/time-unit-math.ts
/**
 * TimeUnitMath — type-safe arithmetic operations on time values.
 * Mirrors the Java TimeUnitMath class.
 *
 * All operations convert through beats as the intermediate representation.
 * Position-type is preserved when adding/subtracting durations.
 */
import { TimePosition } from './time-position';
import { TimeDuration } from './time-duration';
import { TimeBase } from './time-base';
import { TimeContext } from './time-context';

/**
 * Add a duration to a position, returning a new position of the same type.
 */
export function add(context: TimeContext, position: TimePosition, duration: TimeDuration): TimePosition;
/**
 * Add two durations, returning DurationBeats.
 */
export function add(context: TimeContext, a: TimeDuration, b: TimeDuration): TimeDuration;
export function add(context: TimeContext, a: TimePosition | TimeDuration, b: TimeDuration): TimePosition | TimeDuration {
  if (a instanceof TimePosition) {
    // Position + Duration → Position
    const beats = a.toBeats(context) + b.toBeats(context);
    return beatsToTimePosition(Math.max(0, beats), a.getTimeBase(), context);
  } else {
    // Duration + Duration → Duration (beats)
    return TimeDuration.beats(Math.max(0, a.toBeats(context) + b.toBeats(context)));
  }
}

/**
 * Subtract a duration from a position, returning a new position clamped to 0.
 */
export function subtract(context: TimeContext, position: TimePosition, duration: TimeDuration): TimePosition;
/**
 * Subtract two durations, returning DurationBeats clamped to 0.
 */
export function subtract(context: TimeContext, a: TimeDuration, b: TimeDuration): TimeDuration;
export function subtract(context: TimeContext, a: TimePosition | TimeDuration, b: TimeDuration): TimePosition | TimeDuration {
  if (a instanceof TimePosition) {
    // Position - Duration → Position (clamped to 0)
    const beats = Math.max(0, a.toBeats(context) - b.toBeats(context));
    return beatsToTimePosition(beats, a.getTimeBase(), context);
  } else {
    // Duration - Duration → Duration (clamped to 0)
    return TimeDuration.beats(Math.max(0, a.toBeats(context) - b.toBeats(context)));
  }
}

/**
 * Get the absolute distance between two positions as DurationBeats.
 */
export function distance(context: TimeContext, from: TimePosition, to: TimePosition): TimeDuration {
  return TimeDuration.beats(Math.abs(to.toBeats(context) - from.toBeats(context)));
}

/**
 * Get the forward distance (to - from), clamped to 0 if reversed.
 */
export function forwardDistance(context: TimeContext, from: TimePosition, to: TimePosition): TimeDuration {
  return TimeDuration.beats(Math.max(0, to.toBeats(context) - from.toBeats(context)));
}

/**
 * Convert a duration to a different time base.
 * Returns the same instance if the time bases match.
 */
export function convertDuration(dur: TimeDuration, targetBase: TimeBase, context: TimeContext): TimeDuration {
  if (dur.getTimeBase() === targetBase) return dur;
  const beats = dur.toBeats(context);
  return beatsToDuration(beats, targetBase, context);
}

/**
 * Convert raw beats to a duration of the given time base.
 */
export function beatsToDuration(beats: number, targetBase: TimeBase, context: TimeContext): TimeDuration {
  if (beats < 0) beats = 0;

  switch (targetBase) {
    case TimeBase.BEATS:
      return TimeDuration.beats(beats);

    case TimeBase.BBT: {
      const meterMap = context.getMeterMap();
      const meter = meterMap.get(0).meter;
      const beatsPerMeasure = meter.getBeatsPerMeasure();
      const beatScale = meter.getBeatScale();
      const bars = Math.floor(beats / beatsPerMeasure);
      const remaining = beats - bars * beatsPerMeasure;
      const beatNum = Math.floor(remaining / beatScale);
      const fractionalBeat = remaining - beatNum * beatScale;
      const ppq = 960;
      const ticks = Math.round(fractionalBeat * ppq / beatScale);
      return TimeDuration.bbt(bars, beatNum, Math.min(ticks, ppq - 1));
    }

    case TimeBase.BBST: {
      const meterMap = context.getMeterMap();
      const meter = meterMap.get(0).meter;
      const beatsPerMeasure = meter.getBeatsPerMeasure();
      const beatScale = meter.getBeatScale();
      const ppq = 960;
      const bars = Math.floor(beats / beatsPerMeasure);
      const remainingBeats = beats - bars * beatsPerMeasure;
      const scaledBeat = remainingBeats / beatScale;
      const wholeBeat = Math.floor(scaledBeat);
      const fractionalBeat = scaledBeat - wholeBeat;
      const totalTicks = Math.round(fractionalBeat * ppq);
      const ticksPerSixteenth = ppq / 4;
      const sixteenth = Math.floor(totalTicks / ticksPerSixteenth);
      const ticks = totalTicks % ticksPerSixteenth;
      return TimeDuration.bbst(bars, wholeBeat, sixteenth, ticks);
    }

    case TimeBase.BBF: {
      const meterMap = context.getMeterMap();
      const meter = meterMap.get(0).meter;
      const beatsPerMeasure = meter.getBeatsPerMeasure();
      const beatScale = meter.getBeatScale();
      const bars = Math.floor(beats / beatsPerMeasure);
      const remaining = beats - bars * beatsPerMeasure;
      const fullBeats = Math.floor(remaining / beatScale);
      const fractionalBeat = remaining - fullBeats * beatScale;
      const fraction = Math.round(fractionalBeat * 100 / beatScale);
      return TimeDuration.bbf(bars, fullBeats, Math.min(fraction, 99));
    }

    case TimeBase.TIME:
    case TimeBase.SMPTE: {
      const secs = context.getTempoMap().beatsToSeconds(beats);
      return TimeDuration.fromSeconds(secs);
    }

    case TimeBase.SECONDS: {
      const secs = context.getTempoMap().beatsToSeconds(beats);
      return TimeDuration.seconds(secs);
    }

    case TimeBase.FRAME: {
      const secs = context.getTempoMap().beatsToSeconds(beats);
      return TimeDuration.frames(Math.round(secs * context.getSampleRate()));
    }

    default:
      return TimeDuration.beats(beats);
  }
}

/**
 * Convert a TimePosition to a TimeDuration, optionally converting to a target base.
 * The beat value of the position becomes the duration value.
 */
export function fromTimePosition(position: TimePosition, targetBaseOrContext: TimeBase | TimeContext, context?: TimeContext): TimeDuration {
  if (typeof targetBaseOrContext === 'object') {
    // fromTimePosition(position, context) → DurationBeats
    const beats = position.toBeats(targetBaseOrContext);
    return TimeDuration.beats(beats);
  } else {
    // fromTimePosition(position, targetBase, context)
    const beats = position.toBeats(context!);
    return beatsToDuration(beats, targetBaseOrContext, context!);
  }
}

// Re-export beatsToTimePosition from TimeUtilities to avoid circular dependency
// This function is implemented directly here for self-containment
function beatsToTimePositionInternal(beats: number, base: TimeBase, context: TimeContext): TimePosition {
  switch (base) {
    case TimeBase.BEATS:
      return TimePosition.beats(beats);

    case TimeBase.BBT: {
      const meterMap = context.getMeterMap();
      const bbt = meterMap.beatsToBBT(beats);
      return TimePosition.bbt(bbt.bar, bbt.beat, bbt.ticks);
    }

    case TimeBase.BBST: {
      const meterMap = context.getMeterMap();
      const bbst = meterMap.beatsToBBST(beats);
      return TimePosition.bbst(bbst.bar, bbst.beat, bbst.sixteenth, bbst.ticks);
    }

    case TimeBase.BBF: {
      const meterMap = context.getMeterMap();
      const bbf = meterMap.beatsToBBF(beats);
      return TimePosition.bbf(bbf.bar, bbf.beat, bbf.fraction);
    }

    case TimeBase.TIME:
    case TimeBase.SMPTE: {
      const totalSecs = context.getTempoMap().beatsToSeconds(beats);
      const totalMs = Math.round(totalSecs * 1000);
      const hours = Math.floor(totalMs / 3600000);
      const remainingMs = totalMs - hours * 3600000;
      const minutes = Math.floor(remainingMs / 60000);
      const secMs = remainingMs - minutes * 60000;
      const seconds = Math.floor(secMs / 1000);
      const milliseconds = secMs - seconds * 1000;
      return TimePosition.time(hours, minutes, seconds, milliseconds);
    }

    case TimeBase.SECONDS: {
      const secs = context.getTempoMap().beatsToSeconds(beats);
      return TimePosition.seconds(secs);
    }

    case TimeBase.FRAME: {
      const secs = context.getTempoMap().beatsToSeconds(beats);
      return TimePosition.frames(Math.round(secs * context.getSampleRate()));
    }

    default:
      return TimePosition.beats(beats);
  }
}

/** Internal: create position from beats — delegates to beatsToTimePosition. */
function beatsToTimePosition(beats: number, base: TimeBase, context: TimeContext): TimePosition {
  return beatsToTimePositionInternal(beats, base, context);
}
